### PR TITLE
Refactor bottom sheet & related widgets

### PR DESCRIPTION
**Pull request by Pante**
This PR refactors Material's bottom sheet & related widgets. It also replaces internal usage of deprecated types such as `MaterialState` with `WidgetState`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by stuartmorgan** [comment]
test-exempt: code refactor with no semantic change

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 1]
```diff
@@ -431,20 +430,20 @@ class _BottomSheetState extends State<BottomSheet> {
 
 // See scaffold.dart
 
-typedef _SizeChangeCallback<Size> = void Function(Size size);
+typedef _SizeChangeCallback = void Function(Size size);
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/bottom_sheet.dart:433]: The original typedef was quite interesting! 😅

I'm wondering… could we just remove this and use `ValueChanged<Size>`?

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 2]
```diff
@@ -266,11 +265,11 @@ class _BottomSheetState extends State<BottomSheet> {
 
   bool get _dismissUnderway => widget.animationController!.status == AnimationStatus.reverse;
 
-  Set<MaterialState> dragHandleMaterialState = <MaterialState>{};
+  Set<WidgetState> dragHandleState = <WidgetState>{};
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/bottom_sheet.dart:268]: optional nit: Maybe the variable name could reflect how it's a set of WidgetStates.
```suggestion
  Set<WidgetState> dragHandleStates = <WidgetState>{};
```

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 3]
```diff
@@ -431,20 +430,20 @@ class _BottomSheetState extends State<BottomSheet> {
...
-    required this.materialState,
+    required this.state,
     this.dragHandleColor,
     this.dragHandleSize,
   });
 
   final VoidCallback? onSemanticsTap;
   final ValueChanged<bool> handleHover;
-  final Set<MaterialState> materialState;
+  final Set<WidgetState> state;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/bottom_sheet.dart:446]: same nit here:
```suggestion
  final Set<WidgetState> states;
```

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 4]
```diff
@@ -574,50 +573,32 @@ class _RenderBottomSheetLayoutWithSizeListener extends RenderShiftedBox {
...
   @override
   double computeMinIntrinsicWidth(double height) {
-    final double width = _getSize(BoxConstraints.tightForFinite(height: height)).width;
-    if (width.isFinite) {
-      return width;
-    }
-    return 0.0;
+    final double width = BoxConstraints.tightForFinite(height: height).biggest.width;
+    return width.isFinite ? width : 0.0;
   }
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/bottom_sheet.dart:580]: These methods are looking _much_ better now, thanks!

Now that it's been cleaned up a bit, I'm wondering: is it at all possible for this to return anything other than zero? (It seems like `width` will always be set to `double.infinity`, thus `0.0` is always returned.)

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 5]
```diff
@@ -1137,11 +1118,11 @@ class ModalBottomSheetRoute<T> extends PopupRoute<T> {
...
   @override
   Widget buildModalBarrier() {
-    if (barrierColor.alpha != 0 && !offstage) { // changedInternalState is called if barrierColor or offstage updates
-      assert(barrierColor != barrierColor.withOpacity(0.0));
+    if (barrierColor.a != 0 && !offstage) { // changedInternalState is called if barrierColor or offstage updates
+      assert(barrierColor != barrierColor.withValues(alpha: 0.0));
       final Animation<Color?> color = animation!.drive(
         ColorTween(
-          begin: barrierColor.withOpacity(0.0),
+          begin: barrierColor.withValues(alpha: 0.0),
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/bottom_sheet.dart:1125]: Ooh, good catch here!

**Review by nate-thegrate — COMMENTED** [review]
Thanks very much for the cleanup!

File: packages/flutter/lib/src/material/bottom_sheet.dart (hunk 5 above)
Comment (Pante) [reply packages/flutter/lib/src/material/bottom_sheet.dart:1125]: Thanks!

File: packages/flutter/lib/src/material/bottom_sheet.dart (hunk 4 above)
Comment (Pante) [reply packages/flutter/lib/src/material/bottom_sheet.dart]: Oh that's a good catch! It does seem like most of the compute intrinsic methods would always return 0 regardless.

**Review by nate-thegrate — APPROVED** [review]
LGTM, with 1 suggestion :)

Thanks again for the improvement here!

**Review by victorsanni — APPROVED** [review]
LGTM but I don't entirely understand the previous logic in the compute intrinsic height methods.

**Comment by nate-thegrate** [comment]
@victorsanni yeah that's fair.

The original logic was implemented in https://github.com/flutter/flutter/pull/116743; it was a pretty large PR with a focus on improving accessibility.
My guess is that they copy-pasted existing `RenderBox` logic and tweaked it so the box expands to fill the constraints, not realizing that much of the logic relating to intrinsics is unnecessary when [**sizedByParent**](https://main-api.flutter.dev/flutter/rendering/RenderObject/sizedByParent.html) is true.

That PR also added
```dart
typedef _SizeChangeCallback<Size> = void Function(Size);
```
It works exactly like `ValueChanged<T>` but uses `Size` as the type argument, which might be a cause for confusion, given the name overlap with the [**Size**](https://main-api.flutter.dev/flutter/dart-ui/Size-class.html) class.

Overall, I'm super pleased with @Pante's contribution and would happily follow up if anything goes wrong with the merge 👍

**Pull request merged**